src/Features/Lsif/Generator/Generator.cs

```diff
@@ -26,9 +29,9 @@ public Generator(ILsifJsonWriter lsifJsonWriter)
             _lsifJsonWriter = lsifJsonWriter;
         }
 
-        public void GenerateForCompilation(Compilation compilation, string projectPath, HostLanguageServices languageServices)
+        public void GenerateForProject(Project project, Compilation compilation)
```

Comment:
So the lack of Project was intentional here: as @CyrusNajmabadi and I are looking at also making this run as an analyzer, we won't have a Project type to work with, unless we create a wrapper. In that world the code that's doing the per-document stuff would then become the syntax tree/semantic model action so that way we can reuse the caches from the compiler.

Was this necessary to do or can we easily refactor the LSP side of things to have a method that just processes a tree?

Reply:
Alright, if the intent is to move all this down to an analyzer, it makes sense to move the required logic down to `https://github.com/dotnet/roslyn/tree/master/src/Workspaces/SharedUtilitiesAndExtensions/Compiler` and share it with the existing workspace code. The idea that seems hard for me to grasp is all the language services right now are MEF based, and there is no MEF in analyzer layer, so this will be a significant piece of work to even move down to `SharedUtilitiesAndExtensions/Compiler` or consume from an analyzer - we probably need something similar to the infrastructure in CodeStyle analyzer layer, but we will need C# and VB specific LSIF generator projects. I think all of this is a lot of work, so we have below options:

1. Block this PR on all the infrastructure work to move language services based functionality down to non-MEF based `SharedUtilitiesAndExtensions/Compiler` and then implement the outlining functionality in this layer.
2. Go ahead with this PR as-is and have a follow-up PR to do the infra work above, and use Outlining feature with BlockStructureService as the first piece of already implemented functionality down to non-MEF based `SharedUtilitiesAndExtensions/Compiler` layer.

I personally prefer (2), but let me know if you think (1) is better.

Reply:
@jasonmalinowski I put some more thought into this and I will give it a try to do the refactoring in this PR itself. Marking this PR as a draft PR till then.

Reply:
So the LSIF generator already still operates with a MEF container, even if we're only analyzing a Compilation that we created. It just means the only services you can use aren't ones that take Documents/Projects, since that implies a larger universe of information than we have. So you can still use ISyntaxFactsService, you just can't use other things.